client/my-sites/themes/themes-selection.jsx

```diff
@@ -196,13 +199,20 @@ const ConnectedThemesSelection = connect(
 			number
 		};
 
+		const lastQuery = getLastThemeQuery( state, sourceSiteId ) || query;
+		const themesCount = getThemesFoundForQuery( state, sourceSiteId, lastQuery );
+
+		if ( ! isEqual( omit( query, PAGINATION_QUERY_KEYS ), omit( lastQuery, PAGINATION_QUERY_KEYS ) ) ) {
+			query.page = 1;
+		}
+
 		return {
 			query,
 			source: sourceSiteId,
 			siteSlug: getSiteSlug( state, siteId ),
-			themes: getThemesForQueryIgnoringPage( state, sourceSiteId, query ) || [],
-			themesCount: getThemesFoundForQuery( state, sourceSiteId, query ),
-			isRequesting: isRequestingThemesForQuery( state, sourceSiteId, query ),
+			themes: getThemesForQueryIgnoringPage( state, sourceSiteId, lastQuery ) || noThemes,
```

Comment:
Hmm, why do we need to use `lastQuery` here?

This might hint at the fact there's still something a bit wonky about how we handle state here (also see that [duplicated check](https://github.com/Automattic/wp-calypso/pull/12503/files#r107996700)). Looks like we're artificially returning an "older" state instead of handling how/when we update...